Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -192,266 +89,700 @@ void LoadSettings()
+	mapPendingDockingRequests.erase(dockingID);
+
+	auto dockingName = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(dockingID));
+	bool shipAlreadyDocked = false;
+	if (mobiledockClients[carrierID].iDockingModulesAvailable == 0)
+	{
+		for (wstring dockedName : idToCarrierInfoMap[carrierID]->dockedShipList)
 		{
-			// If died carrier and docked ship are in same system.
-			if (Players[client].iSystemID == mobiledockClients[client].carrierSystem)
+			if (dockedName == dockingName)
 			{
-				// Now that the data is prepared, send the player to the carrier's location
-				undockingShip = mobiledockClients[client];
-
-				// Clear the client out of the mobiledockClients now that it's no longer docked
-				mobiledockClients.erase(client);
+				shipAlreadyDocked = true;
+				break;
 			}
-			// If not - redirect to proxy base in carrier's system.
-			else
-			{
-				wstring scProxyBase = HkGetSystemNickByID(mobiledockClients[client].carrierSystem) + L"_proxy_base";
-				if (pub::GetBaseID(mobiledockClients[client].proxyBaseID, (wstos(scProxyBase)).c_str()) == -4)
-				{
-					PrintUserCmdText(client, L"No proxy base in system detected. Contact a developer about this please. Required base: %ws", scProxyBase);
-					return;
-				}
+		}
 
-				mobiledockClients[client].carrierSystem = -1;
-			}
+		if (!shipAlreadyDocked)
+		{
+			PrintUserCmdText(dockingID, L"Carrier has no free docking capacity");
 			return;
 		}
-		else if (carrier_client == -1)
-		{
-			// If the carrier doesn't exist for some reason, force the client to dock on it's last known base
-			uint iBaseID = mobiledockClients[client].iLastBaseID;
-			mobiledockClients[client].undockBase = Universe::get_base(iBaseID);
-			mobiledockClients[client].baseUndock = true;
+	}
 
-			return;
+	// Save the carrier info
+	auto carrierName = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(carrierID));
 
-		}
+	//In case this ship was launched from another mobiledock, unregister it first.
+	RemoveShipFromLists(dockingName, false);
 
-		//Get the carrier ship information
-		uint carrierShip;
-		pub::Player::GetShip(carrier_client, carrierShip);
+	if (!shipAlreadyDocked)
+	{
+		std::lock_guard<std::mutex> saveLock(saveMutex);
+
+		nameToCarrierInfoMap[carrierName].dockedShipList.emplace_back(dockingName);
+		nameToDockedInfoMap[dockingName].lastDockedSolar = Players[dockingID].iLastBaseID;
+		nameToDockedInfoMap[dockingName].carrierName = carrierName;
+		idToCarrierInfoMap[carrierID] = &nameToCarrierInfoMap[carrierName];
+		idToDockedInfoMap[dockingID] = &nameToDockedInfoMap[dockingName];
+		idToCarrierInfoMap[carrierID]->lastCarrierLogin = time(nullptr);
+		mobiledockClients[carrierID].iDockingModulesAvailable--;
+	}
 
-		uint clientShip;
-		pub::Player::GetShip(client, clientShip);
+	// Land the ship on the designated base
+	HkBeamById(dockingID, mobileDockingProxyBase);
+	PrintUserCmdText(carrierID, L"Ship docked");

```

Comment:
Shouldn't we tell the carrier which ship docked?

Reply:
Added